fix: align RoleManager behavior with documentation (#1404)

bugoutianzhen123 opened the pull request:
This PR introduces a level field to the Role struct and updates the RoleManager to dynamically maintain this field.
It ensures that maxHierarchyLevel not only limits the depth of the HasLink function, but also applies to other role traversal  such as:
GetImplicitRolesForUser
GetImplicitUsersForRole

Why this change is necessary
Previously, maxHierarchyLevel only affected HasLink, while other functions could traverse indefinitely, which led to inconsistencies between the documented behavior and actual implementation.
This update:
Makes the behavior of RoleManager consistent with the documentation
Keeps the change self-contained within RoleManager, without requiring modifications to external components
Ensures all relevant traversal functions now honor the maxHierarchyLevel setting

Summary of Changes
Added level field to Role struct
Dynamically maintained level values within RoleManager
Applied hierarchy level checks consistently across relevant APIs

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/casbin/casbin?pullRequest=1499) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/casbin/casbin?pullRequest=1499) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/casbin/casbin?pullRequest=1499) it.</sub>

hsluoyz commented:
@tangyang9464 plz review

tangyang9464 commented:
@bugoutianzhen123 Tests need to be added

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -46,12 +48,41 @@ func newRole(name string) *Role {
+		})
+	}
+}
+
```
Consider streamlining redundant local variables and overall processes
example:
```go
func (r *Role) updateLevel() {
	maxLevel := 0
	r.users.Range(func(_, value interface{}) bool {
		if parentRole := value.(*Role); parentRole.level >= 0 && parentRole.level+1 > maxLevel {
			maxLevel = parentRole.level + 1
		}
		return true
	})
	if r.level == maxLevel {
		return
	}
	r.level = maxLevel
	r.roles.Range(func(_, value interface{}) bool {
		value.(*Role).updateLevel()
		return true
	})
}
```

Also, what if the role map has a ring?

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -331,6 +362,9 @@ func (rm *RoleManagerImpl) Clear() error {
 	user, _ := rm.getRole(name1)
 	role, _ := rm.getRole(name2)
+	if rm.maxHierarchyLevel > 0 && role.level >= rm.maxHierarchyLevel-1 {
+		return errors.New("exceeded level limit")
```
Need more detailed error message

bugoutianzhen123 commented:
Thank you for the previous review!
I’ve updated the PR with the following improvements:

Added cycle detection to prevent circular role inheritance;
Added corresponding test cases;
Optimized some internal logic for better readability and flow;
Improved error messages for better debugging;
Also, I noticed that the TestMaxHierarchyLevel test previously assumed that the setting limits the "difference in role levels." However, based on the documentation and common usage, maxHierarchyLevel should limit the length of the inheritance chain. I've adjusted the test and logic accordingly.

Please take another look when convenient

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -46,17 +47,40 @@ func newRole(name string) *Role {
 	r.roles.Store(role.name, role)
 	role.addUser(r)
+	r.updateLevel()
+	role.updateLevel()
```
Why is there an extra line? It seems to be repeated.
In method `updateLevel`
```go
r.roles.Range(func(_, value interface{}) bool {
		value.(*Role).updateLevel()
		return true
	})
```

bugoutianzhen123 replied:
This approach ensures that when two inheritance chains are merged into one, the levels can be correctly updated. I will try to optimize it.

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -329,9 +355,21 @@ func (rm *RoleManagerImpl) Clear() error {
+		return err
+	} else if hasLink {
+		return fmt.Errorf("cannot add link from '%s' to '%s': would create a cycle", name1, name2)
+	}
```
It is best not to use `HasLink` for loop detection, because it not only includes normal role inheritance, but also pattern matching, which does not match the logic of method `updateLevel`. I think it is better to use `visit[]` to mark in updateLevel

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -329,9 +355,21 @@ func (rm *RoleManagerImpl) Clear() error {
+		user.removeRole(role)
+		return fmt.Errorf("cannot add link from '%s' to '%s': role '%s' is at level %d, exceeding the maximum allowed hierarchy level of %d",
+			name1, name2, name2, role.level, rm.maxHierarchyLevel)
+	}
```
Can it be detected before `addRole`?

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -392,7 +430,8 @@ func (rm *RoleManagerImpl) GetRoles(name string, domains ...string) ([]string, e
 	}
-	return user.getRoles(), nil
+
+	return rm.filterHelper(user.getRoles()), nil
```
Since the `level` has been restricted in `AddLink`, will `getRoles` still exceed the level? If not, is filtering unnecessary?

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -228,7 +252,9 @@ func (rm *RoleManagerImpl) rebuild() {
 	rangeLinks(roles, func(name1, name2 string, domain ...string) bool {
-		_ = rm.AddLink(name1, name2, domain...)
+		if err := rm.AddLink(name1, name2, domain...); err != nil {
+			rm.logger.LogError(err)
```
This logic has little to do with PR, just keep it as it is

tangyang9464 commented on rbac/default-role-manager/role_manager.go:
```diff
@@ -402,7 +441,37 @@ func (rm *RoleManagerImpl) GetUsers(name string, domain ...string) ([]string, er
+}
+
+func (rm *RoleManagerImpl) filterHelper(names []string) []string {
+	if rm.maxHierarchyLevel <= 0 {
```
Why not just modify role.getUsers()

tangyang9464 commented on rbac/default-role-manager/role_manager_test.go:
```diff
@@ -364,29 +364,47 @@ func TestTemporaryRoles(t *testing.T) {
 }
 
 func TestMaxHierarchyLevel(t *testing.T) {
-	rm := NewRoleManager(1)
```
To maintain compatibility, I think it is best not to restrict AddLink by level, and only filter by level when gettingRoles/Uses and HasLink.

tangyang9464 commented:
> TestMaxHierarchyLevel test previously assumed that the setting limits the "difference in role levels." However, based on the documentation and common usage, maxHierarchyLevel should limit the length of the inheritance chain. I've adjusted the test and logic accordingly.

In order to maintain compatibility, I think we only need to restrict `getRoles/Users`, `HasLink`, etc. For users, they can flexibly adjust maxLevel

bugoutianzhen123 commented:
I believe it's better to separate concerns: AddLink should handle cycle detection, while updateLevel should be dedicated to maintaining the level structure. Level-based filtering should only be enforced in role resolution methods like getRoles, getUsers, and HasLink

bugoutianzhen123 commented on rbac/default-role-manager/role_manager_test.go:
```diff
@@ -364,29 +364,47 @@ func TestTemporaryRoles(t *testing.T) {
+	testRole(t, rm2, "level0", "level3", false)
+	testRole(t, rm2, "level1", "level2", true)
+	testRole(t, rm2, "level1", "level3", false)
+	testRole(t, rm2, "level2", "level3", false)
```
The old TestMaxHierarchyLevel test previously assumed that the setting limits the "difference in role levels." However, the  new TestMaxHierarchyLevel think maxHierarchyLevel should limit the length of the inheritance chain.

hsluoyz commented:
I don't think this code will be OK if merged given the code quality shown here. Closed for now.
